Implement Shelly 3EM

Sn0w3y opened the pull request:
Link to Shelly [Product](https://www.shelly.com/de/products/shop/shelly-3em-1)

This is a 3-phase Meter, which can be used as a Grid Meter for example.

sfeilmeier commented:
@michaelgrill Please check usage of HTTP-Bridge

michaelgrill commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3EMImpl.java:
```diff
@@ -0,0 +1,270 @@
+import io.openems.edge.meter.api.SinglePhaseMeter;
+import io.openems.edge.timedata.api.Timedata;
+import io.openems.edge.timedata.api.TimedataProvider;
+import io.openems.edge.timedata.api.utils.CalculateEnergyFromPower;
```
i get errors here. You probably need to add `io.openems.edge.timedata.api` to your bnd file

michaelgrill commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3EMImpl.java:
```diff
@@ -0,0 +1,270 @@
+	        JsonObject updateObject = jsonResponse.getAsJsonObject("update");
+	        if (updateObject != null) {
+	            boolean hasUpdate = updateObject.get("has_update").getAsBoolean();
+				this.channel(IoShellyPlus3EM.ChannelId.HAS_UPDATE).setNextValue(hasUpdate);
```
this value does not exist you need to add it in the enum

michaelgrill commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3EMImpl.java:
```diff
@@ -0,0 +1,270 @@
+	            JsonObject emeter = emeters.get(i).getAsJsonObject();
+	            Float power = emeter.get("power").getAsFloat();
+	            Float voltage = emeter.get("voltage").getAsFloat();
+	            Float current = emeter.get("current").getAsFloat();
```
please use primitives there is a difference between the type `float` and `Float` and its better to use primitives most of the times. You can read more about it [here](https://www.baeldung.com/java-primitives-vs-objects)

sfeilmeier replied:
I also always recommend to use the appropriate JsonUtils methods, because they handle Exceptions better.

michaelgrill commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3EMImpl.java:
```diff
@@ -0,0 +1,270 @@
+	private void setValuesForPhase(int phase, Float voltage, Float current, Float power) {
+	    Integer scaledVoltage = Math.round(voltage * 1000);
+	    Integer scaledCurrent = Math.round(current * 1000);
+	    Integer scaledPower = Math.round(power);
```
same thing with primitives

michaelgrill reviewed (changes requested):
General usage of HttpBridge is fine. Had some other things i noticed and also format your code 🙂 (CTRL + SHIFT + F in Eclipse)

sfeilmeier commented:
> @sfeilmeier the other Error on Test was because of AbbreviationAsWordInName - the name is 3EM - i would be glad if we could keep this as it is as this is the original name - what do you think?

It's very often like this unfortunately. It's possible to ignore the Checkstyle error (using `// CHECKSTYLE:OFF` - and `:ON`, but generally you should rename it to e.g. `3em`

Sn0w3y commented:
Done ! @michaelgrill  :)

michaelgrill commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java:
```diff
@@ -0,0 +1,266 @@
+
+	private volatile Timedata timedata;
+    
+    @Reference(cardinality = ReferenceCardinality.MANDATORY)
```
`ReferenceCardinality.MANDATORY` is the default one you can omit it if you want

michaelgrill commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java:
```diff
@@ -0,0 +1,266 @@
+
+    @Deactivate
+    protected void deactivate() {
+        if (this.httpBridge != null) {
```
you dont need to check if the `httpBridge` is null the `unget` method allows null values and ignores them

michaelgrill reviewed (changes requested):
usage in general looks fine only some side notes and did formatting change again? 🤔
after the merge of https://github.com/OpenEMS/openems/pull/2539 we should be able to merge this one as well

Sn0w3y commented:
@michaelgrill should be done now ! :) ready to check and review again :)

If your new HttpBridge is merged it would be great if this also would be merged ? :)

Sn0w3y commented:
as the HttpBridge works now maybe up ?

michaelgrill reviewed (approved):
Looks good to me. 👍 ty
(i was waiting for https://github.com/OpenEMS/openems/pull/2539#pullrequestreview-1894547131 to be merged)

sfeilmeier commented on io.openems.edge.io.shelly/bnd.bnd:
```diff
@@ -10,6 +10,6 @@ Bundle-Version: 1.0.0.${tstamp}
 	io.openems.edge.io.api,\
 	io.openems.edge.meter.api,\
-
+	io.openems.edge.timedata.api
```
I am changing this to `	io.openems.edge.timedata.api,\` and adding a newline. This is because prepare-commit.sh sorts the entries alphabetically. A missing trailing "\" can then possibly break the build.

Sn0w3y replied:
sure, understood

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/Config.java:
```diff
@@ -0,0 +1,29 @@
+import io.openems.edge.meter.api.MeterType;
+
+@ObjectClassDefinition(//
+		name = "IO Shelly Plus3EM", //
```
Official name seems to be `Shelly 3EM`. I am renaming everything (classes, package, etc.)

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3em.java:
```diff
@@ -0,0 +1,146 @@
+		 * <li>Level: INFO
+		 * </ul>
+		 */
+		HAS_UPDATE(Doc.of(Level.INFO) //
```
This is an interesting feature. Should we standardize this Channel for alle Shellys? - i.e. create a `Shelly` nature that provides this information and a generic handler, that parses the `has_update` value?

Sn0w3y replied:
Yes, can be done, but not all shellys provide this info as a boolean. some of them just show the new versions in the responses for example ShellyPlus1PM:

`{"ble":{},"cloud":{"connected":true},"input:0":{"id":0,"state":false},"mqtt":{"connected":false},"switch:0":{"id":0, "source":"WS_in", "output":false, "apower":0.0, "voltage":230.2, "current":0.000, "aenergy":{"total":8629.000,"by_minute":[0.000,0.000,0.000],"minute_ts":1708707240},"temperature":{"tC":43.5, "tF":110.3}},"sys":{"mac":"80646FE34998","restart_required":false,"time":"17:54","unixtime":1708707267,"uptime":491734,"ram_size":260248,"ram_free":108996,"fs_size":458752,"fs_free":143360,"cfg_rev":21,"kvs_rev":2,"schedule_rev":0,"webhook_rev":0,"**available_updates":{"stable":{"version":"1.2.1"}}**,"reset_reason":3},"wifi":{"sta_ip":"192.168.178.169","status":"got ip","ssid":"heizung","rssi":-50,"ap_client_count":0},"ws":{"connected":false}}`

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java:
```diff
@@ -0,0 +1,249 @@
+	private void processHttpResult(JsonElement result, Throwable error) {
+		this._setSlaveCommunicationFailed(result == null);
+		if (error != null) {
+			this._setRelay(null);
```
Out of experience I recommend to collect all values first and then set the Channel values all together. This way you will not forget to set a channel in any case. 

For example here you forget to (un)set `hasUpdate` and individual phase values in case of an error

Sn0w3y replied:
thanks !

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java:
```diff
@@ -0,0 +1,249 @@
+			if (relays != null && !relays.isEmpty()) {
+				JsonObject relay = JsonUtils.getAsJsonObject(relays.get(0));
+				boolean isOn = JsonUtils.getAsBoolean(relay, "ison");
+				this.getRelayChannel().setNextWriteValue(isOn);
```
Why are you setting the write-value here? Sounds very wrong. I'll remove it, but please correct me if I am wrong.

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java:
```diff
@@ -0,0 +1,249 @@
+			JsonObject jsonResponse = JsonUtils.getAsJsonObject(result);
+
+			JsonArray relays = JsonUtils.getAsJsonArray(jsonResponse, "relays");
+			if (relays != null && !relays.isEmpty()) {
```
`relays` can never be `null` here. `getAsJsonArray` would throw an exception otherwise

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplus3em/IoShellyPlus3emImpl.java:
```diff
@@ -0,0 +1,249 @@
+@Component(name = "IO.Shelly.Plus3EM", immediate = true, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@EventTopics({ EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE, EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE })
+public class IoShellyPlus3emImpl extends AbstractOpenemsComponent implements IoShellyPlus3em, DigitalOutput,
+		SinglePhaseMeter, ElectricityMeter, OpenemsComponent, TimedataProvider, EventHandler {
```
Why is this a SinglePhaseMeter? Anyway you forget to add the Channels in constructor, so a JUnit test would have prohibited this. I'll a a JUnit test and remove this here.

Sn0w3y replied:
thanks !

sfeilmeier reviewed (commented):
Please find my comments. I applied all suggested changes. Can you please validate that everything still works?

Sn0w3y commented:
I will test this and report if it works :) @sfeilmeier

Sn0w3y commented:
@sfeilmeier should be completed now ! I just added some functionalities - maybe just have a little look after it:

this.channel(IoShelly3Em.ChannelId.EMETER1_EXCEPTION).setNextValue(!isValid);
this.channel(IoShelly3Em.ChannelId.EMETER2_EXCEPTION).setNextValue(!isValid);
this.channel(IoShelly3Em.ChannelId.EMETER3_EXCEPTION).setNextValue(!isValid);


also added:

this.channel(IoShelly3Em.ChannelId.RELAY_OVERPOWER_EXCEPTION).setNextValue(overpower);


to see if the Relais is overpowered :)

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly3em/IoShelly3EmImpl.java:
```diff
@@ -160,34 +161,43 @@ private void processHttpResult(JsonElement result, Throwable error) {
+					var current = round(getAsFloat(emeter, "current") * 1000);
+					var isValid = getAsBoolean(emeter, "is_valid");
+
+					activePower = round(totalPower);
```
This could be before the for-loop, right?

Just to `activePower = getAsFloat(response, "total_power");` in line 172

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly3em/IoShelly3EmImpl.java:
```diff
@@ -150,6 +150,7 @@ private void processHttpResult(JsonElement result, Throwable error) {
 		Integer currentL2 = null;
 		Integer currentL3 = null;
 		Boolean hasUpdate = false;
+		Boolean overpower = null;
```
StateChannels should be `boolean` instead of `Boolean` (this was wrong with my hasUpdate). Pre-initialize with `false`.

Sn0w3y commented:
fixed and working :D @sfeilmeier

sfeilmeier commented on io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly3em/IoShelly3EmImpl.java:
```diff
@@ -0,0 +1,283 @@
+		}
+	}
+
+	private void processHttpResult(JsonElement result, Throwable error) {
```
For future PRs, please provide an example Json result. Then we can write proper JUnit tests. Right now this code has only 33 % code coverage.

Sn0w3y replied:
Example JSON:

{"wifi_sta":{"connected":true,"ssid":"Fritzlbox","ip":"192.168.178.130","rssi":-41},"cloud":{"enabled":true,"connected":true},"mqtt":{"connected":false},"time":"11:31","unixtime":1708770669,"serial":5,"has_update":false,"mac":"485519D6A289","cfg_changed_cnt":0,"actions_stats":{"skipped":0},"relays":[{"ison":false,"has_timer":false,"timer_started":0,"timer_duration":0,"timer_remaining":0,"overpower":false,"is_valid":true,"source":"http"}],"emeters":[{"power":3.81,"pf":0.33,"current":0.05,"voltage":231.38,"is_valid":true,"total":0.6,"total_returned":0.0},{"power":3.86,"pf":0.33,"current":0.05,"voltage":231.36,"is_valid":true,"total":0.6,"total_returned":0.0},{"power":3.83,"pf":0.33,"current":0.05,"voltage":231.36,"is_valid":true,"total":0.6,"total_returned":0.0}],"total_power":11.50,"emeter_n":{"current":0.00,"ixsum":0.15,"mismatch":false,"is_valid":false},"fs_mounted":true,"v_data":1,"ct_calst":0,"update":{"status":"idle","has_update":false,"new_version":"20230913-114244/v1.14.0-gcb84623","old_version":"20230913-114244/v1.14.0-gcb84623","beta_version":"20231107-165007/v1.14.1-rc1-g0617c15"},"ram_total":49920,"ram_free":29108,"fs_size":233681,"fs_free":155118,"uptime":251}
